File: PrimePY_2_jgp.pl
#!/usr/bin/env perl

use strict;
use warnings;

package PrimeSieve {

    my %DICT = (
        10          => 4,
        60          => 17,
        100         => 25,
        1000        => 168,
        10000       => 1229,
        100000      => 9592,
        1000000     => 78498,
        10000000    => 664579,
        100000000   => 5761455,
        1000000000  => 50847534,
        10000000000 => 455052511,
    );

    sub new {
        my ( $class, $sieve_size, $use_inline ) = @_;

        # Each byte of $bytes represents on odd number in our virtual
        # bit map; i.e. positions 1,3,5...size/2
        # This is so we can ignore even numbers, which except for 2
        # (a special case) are all non-prime.
        # We start by initializing this bytemap to 1's to indicate
        # all positions are non-prime; then we'll set primes to 0
        my $bytes = '1' x int(($sieve_size+1)/2);

        bless {
            sieve_size => $sieve_size,
            bytes       => $bytes,
            iloops      => 0,
            mloops      => 0,
            inline      => $use_inline,
        }, $class;
    }

    sub run_sieve {
        my $self     = shift;

        my $inline   = $self->{inline};
        my $size     = $self->{sieve_size};
        my $bytes    = \$self->{bytes};
        my $byteslen = length $$bytes;
        my $mloops   = 0;
        my $iloops   = 0;
        my $q        = sqrt($size) / 2;
        my $factor   = 1;

        #printf "bytes %s\n", $$bytes;

        #                 sqrt(60)->|
        # pos    0 1 2 3|4  5  6  7 |  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29
        # factor 1 3 5 7|9 11 13 15 | 17 19 21 23 25 27 29 31 33 35 37 39 41 43 45 47 49 51 53 55 57 59
        #        1 1 1 1|1  1  1  1 |  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  1  0  1  1  0
        #               q

        while ( $factor <= $q ) {
            # factor will be 3, 5, 7 ... $q
            $factor = index($$bytes, '1', $factor);

            # If marking factor 3, you wouldn't mark 6 (it's a mult of 2)
            # so start with the 3rd instance of this factor's multiple.
            # We can then step by factor * 2 because every second one
            # is going to be even by definition.

            my $start = 2 * $factor * ($factor + 1);
            my $step = $factor * 2 + 1;

            if ($inline) {
                $self->{mloops} += main::set_rng_inline($$bytes, $start, $byteslen, $step, '0');
            } else {
                $self->{iloops} += $self->set_rng($start, $byteslen, $step, '0');
            }
            $self->{iloops}++;  # count this loop too

            $factor++;
            # printf "bytes %s\n", $$bytes;
        }
    }

    sub set_rng {
        my ($self, $from, $to, $step, $v) = @_;

        my $bytes = \$self->{bytes};
        my $cnt = 0;

        for (my $i = $from; $i <= $to; $i += $step) {
            # print "offset = $i byteslen = $to step = $step \n";
            substr($$bytes, $i, 1) = $v;
            $cnt++;
        }
        return $cnt;
    }

    sub get_primes {
        my $self = shift;
        my $bytes = \$self->{bytes};
        my $size  = $self->{sieve_size};

        my $idx = 1;
        my @primes;

        # Since we auto-filter evens, we have to special case the number 2 which is prime
        push @primes, 2 if ($size > 1);

        return @primes unless $size > 2;

        while ($idx > 0) {
            push @primes, $idx * 2 + 1;
            $idx = index($$bytes,'1', $idx + 1);
        }

        return @primes;
        # First, construct a list consisting of 2 and then all odd numbers
        # from 3 to the sieve size.  Then select those entries that
        # have a corresponding bit marked 0 in the bit map.
        # The attempted performance improvement here is to use grep
        # to avoid the overhead of an interpreter loop (e.g. while or for)
        # though profiling indicates it's not a win.

        # grep !substr($$bytes,$_,1), (2, grep $_ % 2, 3..$self->{sieve_size} );
    }

    sub print_results {
        my ( $self, $show_primes, $duration, $passes ) = @_;
        my @primes = $self->get_primes();
        my $count = @primes;
        my $f = $self->validate_results($count);
        printf "%s\n", join ',', @primes if $show_primes;
        my $script = (split /\\/, $0)[-1];    # get script name sans path
        printf "jgpuckering/$script;%d;%f;%d;algorithm=base,faithful=%s,bits=8\n", $passes, $duration, 1, $f;
        printf {*STDERR} "Passes: %d, Time: %.2f, Avg: %f, Passes/sec: %.1f, Limit: %d, Count: %d, Valid: %s\n",
          $passes, $duration, $duration / $passes, $passes / $duration, 
          $self->{sieve_size}, $count, $f;
    }

    sub validate_results {
        my ($self, $count) = @_;

        my $f
            = $DICT{$self->{sieve_size}}
            ? $DICT{$self->{sieve_size}} == $count ? 'yes' : 'no'
            : 'unknown';
    }
};

package main;

use Getopt::Long;
use Time::HiRes 'time';

use Inline C => <<~'__C__';
    int set_rng_inline(char *str, int start, int stop, int step, char *v) {
        int i;
        int cnt = 0;

        for (i=start; i<stop; i+=step) {
            str[i] = v[0];
            cnt++;
        }
        return cnt;
    }
__C__

my $opt_size      = 1_000_000;
my $opt_passes    = 0;
my $opt_primes    = 0;
my $opt_nostats   = 0;
my $opt_duration  = 5;
my $opt_inline    = 0;

GetOptions(
    'size=i',    => \$opt_size,         # sieve size
    'passes=i'   => \$opt_passes,       # by passes not duration
    'primes'     => \$opt_primes,       # show primes
    'nostats'    => \$opt_nostats,      # don't show stats
    'inline'     => \$opt_inline,       # use inline set_rng
);

my $mloops   = 0;
my $iloops   = 0;
my $passes   = 0;
my $duration = 0;
my $sieve;

sub duration {
    my $start_time = shift;
    time() - $start_time;
}

my $start_time = time;

if ($opt_passes) {
    for ( my $i = 0; $i < $opt_passes; $i++ ) {
        $sieve = PrimeSieve->new($opt_size, $opt_inline);
        $sieve->run_sieve();
        $mloops += $sieve->{mloops};
        $iloops += $sieve->{iloops};
        $passes++;
    }
} else {
    while (duration($start_time) < $opt_duration ) {
        $sieve = PrimeSieve->new($opt_size, $opt_inline);
        $sieve->run_sieve();
        $mloops += $sieve->{mloops};
        $iloops += $sieve->{iloops};
        $passes++;
    }
}

$sieve->print_results( $opt_primes, duration($start_time), $passes )
    unless $opt_nostats;

my $rng_sub = $opt_inline ? '' : '';
printf("interpreter loops = $iloops  machine loops = $mloops\n");

__END__

File: PrimePY_2_jgp.py
"""
Python Prime Sieve

MyFirstPython Program (tm) Dave Plummer 8/9/2018
Updated 3/22/2021 for Dave's Garage episode comparing C++, C#, and Python
"""

from math import sqrt


class PrimeSieve:

    """This is the main PrimeSieve class. Call it with the number you wish as
    an upper limit, then call the run_sieve method to do the calculation.
    print_results will dump the count to check validity."""

    prime_counts = { 10 : 4,                 # Historical data for validating our results - the number of primes
                     40 : 12,
                     60 : 17,
                    100 : 25,                # to be found under some limit, such as 168 primes under 1000
                    1000 : 168,
                    10000 : 1229,
                    100000 : 9592,
                    1000000 : 78498,
                    10000000 : 664579,
                    100000000 : 5761455
                  }

    def __init__(self, limit):
        self._size = limit
        self._bits = bytearray(b"\x01") * ((self._size + 1) // 2)
        self._mloops = 0
        self._iloops = 0

    def validate_results(self):                      # Check to see if this is an upper_limit we can

        """Look up our count of primes in the historical data (if we have it)
        to see if it matches"""

        if self._size in self.prime_counts:                              # the data, and (b) our count matches. Since it will return
            return self.prime_counts[self._size] == self.count_primes()  # false for an unknown upper_limit, can't assume false == bad
        return False

    def run_sieve(self, forloop):

        """Calculate the primes up to the specified limit"""

        factor = 1
        # sqrt doesn't seem to make any difference in CPython,
        # but works much faster than "x**.5" in Pypy
        q = sqrt(self._size) / 2
        bitslen = len(self._bits)


        while factor <= q:
            factor = self._bits.index(b"\x01", factor)

            # If marking factor 3, you wouldn't mark 6 (it's a mult of 2) so start with the 3rd instance of this factor's multiple.
            # We can then step by factor * 2 because every second one is going to be even by definition
            start = 2 * factor * (factor + 1)
            step  = factor * 2 + 1
            size  = bitslen - start
            # print("size %d bitslen %d start %d step %d size %d" % (self._size, bitslen, start, step, size))
            # print(b"\x00" * (size // step + bool(size % step)) )
            times = size // step + bool(size % step)    # bool is (a subclass of) int in python

            if forloop:
                # use for loop with step
                for i in range(start, bitslen, step):
                    # print("i", i, "bits[i:] before", self._bits[i:i+1])
                    self._bits[i:i+1] = b"\x00";            
                    # print("i", i, "bits[i:]  after", self._bits[i:i+1])
                    self._iloops += 1
                # print(self._bits)
            else:
                # use slicing with step
                self._bits[start :: step] = b"\x00" * times
                self._mloops += times
                self._iloops += 1
            
            factor += 1

    def count_primes(self):

        """Return the count of bits that are still set in the sieve.
        Assumes you've already called run_sieve, of course!"""

        return self._bits.count(b"\x01") if self._size > 1 else 0

    def get_primes(self):

        """Returns a generator to iterate over the found prime numbers.
        Requires a prior run_sieve call"""

        if self._size > 1:
            yield 2  # Since we auto-filter evens, we have to special case the number 2 which is prime
        if self._size > 2:
            num = 1
            while num > 0:
                yield num * 2 + 1
                num = self._bits.find(1, num + 1)

    def print_results(self, show_results, duration, passes):

        """Displays the primes found (or just the total count,
        depending on what you ask for)"""

        count = 0
        for num in self.get_primes():  # Count (and optionally dump) the primes that were found below the limit
            count += 1
            if show_results:
                print("%s, " % num, end="")
        if show_results:
            print()

        # Following 2 lines added by rbergen to conform to drag race output format
        print("ssovest(jgp); %d;%6f;1;algorithm=base,faithful=yes,bits=8" % (passes, duration));

        print("Passes: %s, Time: %.2f, Avg: %f, Passes/sec: %.1f, Limit: %s, Count: %s, Valid: %s" \
            % ( passes, duration, duration/passes, passes/duration, self._size, count, self.validate_results() ), \
            file=stderr)



# MAIN Entry
if __name__ == "__main__":
    from argparse import ArgumentParser
    from timeit import default_timer  # For timing the durations

    parser = ArgumentParser(description="Python Prime Sieve")
    parser.add_argument("--passes", "-p", help="Number of passes", type=int, default=0)
    parser.add_argument("--limit",  "-l", help="Upper limit for calculating prime numbers", type=int, default=1_000_000)
    parser.add_argument("--time",   "-t", help="Time limit", type=float, default=5)
    parser.add_argument("--show",   "-s", help="Print found prime numbers", action="store_true")
    parser.add_argument("--forloop","-f", help="Use for loop instead of slicer", action="store_true")

    args         = parser.parse_args()
    limit        = args.limit
    timeout      = args.time
    show_results = args.show
    passes       = args.passes
    forloop      = args.forloop

    time_start = default_timer()                           # Record our starting time
    
    if passes > 0:
        for n in range(0, passes):
            sieve = PrimeSieve(limit)                          # Calc the primes up to a million
            sieve.run_sieve(forloop)                           # Find the results
    else:
        time_start = default_timer()                           # Record our starting time
        while (default_timer() - time_start < timeout):        # Run until more than 10 seconds have elapsed
            sieve = PrimeSieve(limit)                          # Calc the primes up to a million
            sieve.run_sieve(forloop)                           # Find the results
            passes = passes + 1                                # Count this pass

    time_delta = default_timer() - time_start              # After the "at least 10 seconds", get the actual elapsed
    sieve.print_results(show_results, time_delta, passes)  # Display outcome
    print("interpreter loops = %s  est machine loops = %s" % (sieve._iloops, sieve._mloops));

